lib/src/clients/relay_client.dart

```diff
@@ -70,53 +73,93 @@ class RelayClient implements UhstRelayClient {
   }) async {
     final hostUrl = sendUrl ?? relayUrl;
     final uri = Uri.parse('$hostUrl?token=$token');
-    http.Response response;
     try {
-      response = await http.post(
+      await http.post(
         uri,
         headers: <String, String>{
           _RelayClientConsts.requestHeaderContentName:
               _RelayClientConsts.requestHeaderContentValue,
         },
         body: message,
       );
-    } on Exception catch (exception) {
-      throw RelayUnreachable(exception);
-    }
-    switch (response.statusCode) {
-      case 200:
-        return;
-      case 400:
-        throw InvalidClientOrHostId(response.body);
-      case 401:
-        throw InvalidToken(token);
-      default:
-        throw RelayException(response.body);
+    } on Exception catch (e) {
+      if (e is NetworkException) {
+        switch (e.responseCode) {
+          case 400:
+            throw InvalidClientOrHostId(e.message);
+          case 401:
+            throw InvalidToken(e.message);
+          default:
+            throw RelayUnreachable(e.message);
+        }
+      } else {
+        throw RelayUnreachable(e);
+      }
     }
   }
 
   @override
-  void subscribeToMessages({
+  Future<void> subscribeToMessages({
     required String token,
     required RelayReadyHandler onReady,
     required RelayExceptionHandler onException,
     required RelayMessageHandler onMessage,
+    RelayEventHandler? onRelayEvent,
     String? receiveUrl,
-  }) {
+  }) async {
     final url = receiveUrl ?? relayUrl;
     final finalUrl = '$url?token=$token';
+    final completer = Completer();
 
-    final EventSource source = EventSource(finalUrl);
-    source.onOpen.listen((event) {
-      onReady(stream: RelayStream(eventSource: source));
-    });
-    source.onError.listen((event) {
-      onException(exception: RelayException(event));
-    });
-    source.onMessage.listen((event) {
-      final eventMessageMap = jsonDecode(event.data);
-      final eventMessage = EventMessage.fromJson(eventMessageMap);
-      onMessage(message: eventMessage.body);
-    });
+    void onCompleterNotResolved(
+      VoidCallback callback, {
```

Comment:
why do we need this method? The logic here is the following:
- if EventSource constructor fails and not completer.isComplete then completer.completeError
- else if completer.isComplete or onError call onException
That's it, I am confused why we have VoidCallback and onCompleterNotResolved

Reply:
it is a method inside a method just for convenience no to write it
onError may come more than once and the first time it will completer.completeError
and after the second all onError will be call callback function but not completer.completeError

Reply:
But it's much simpler to just check completer.isComplete, please see the corresponding JS code (there is no isComplete in JS so had to track with a flag): 
`
let resolved = false;
      const stream = new EventSource(`${url}?token=${token}`);
      stream.onopen = () => {
        if (!resolved) {
          resolve(stream);
          resolved = true;
        }
      };
      stream.onerror = () => {
        if (!resolved) {
          // error on connect
          reject(new RelayError());
          resolved = true;
        } else if (relayErrorHandler) {
          relayErrorHandler(new RelayError());
        }
      };
`

Reply:
will fix it

Reply:
fixed